### PR TITLE
netbox: 3.7.8 -> 4.1.3

### PR DESCRIPTION
**Pull request by minijackson**
## Description of changes

Finally!

Some upgrades and new Python packages were needed by the NetBox upgrade.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - NixOS tests: `netbox` and `netbox-upgrade`
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [x] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

cc NetBox maintainers: @RaitoBezarius & @n0emis 
django-mptt maintainer: @mweinelt
and people interested in the upgrade: @KiruyaMomochi & @xanderio & @Xyz00777

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by xanderio** [comment]
Result of `nixpkgs-review pr 331750` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
  </ul>
</details>
<details>
  <summary>5 packages failed to build:</summary>
  <ul>
    <li>netbox</li>
    <li>python311Packages.netbox-reorder-rack</li>
    <li>python311Packages.netbox-reorder-rack.dist</li>
    <li>python312Packages.netbox-reorder-rack</li>
    <li>python312Packages.netbox-reorder-rack.dist</li>
  </ul>
</details>
<details>
  <summary>16 packages built:</summary>
  <ul>
    <li>python311Packages.django-choices-field</li>
    <li>python311Packages.django-choices-field.dist</li>
    <li>python311Packages.django-htmx</li>
    <li>python311Packages.django-htmx.dist</li>
    <li>python311Packages.django-mptt</li>
    <li>python311Packages.django-mptt.dist</li>
    <li>python311Packages.strawberry-django</li>
    <li>python311Packages.strawberry-django.dist</li>
    <li>python312Packages.django-choices-field</li>
    <li>python312Packages.django-choices-field.dist</li>
    <li>python312Packages.django-htmx</li>
    <li>python312Packages.django-htmx.dist</li>
    <li>python312Packages.django-mptt</li>
    <li>python312Packages.django-mptt.dist</li>
    <li>python312Packages.strawberry-django</li>
    <li>python312Packages.strawberry-django.dist</li>
  </ul>
</details>

File: pkgs/development/python-modules/django-mptt/default.nix [hunk 1]
```diff
@@ -2,35 +2,33 @@
...
   ];
 
   pythonImportsCheck = [ "mptt" ];
 
-  checkPhase = ''
-    runHook preCheck
-    ${python.interpreter} tests/manage.py test
-    runHook postCheck
-  '';
+  # No pytest checks, since they depend on model_mommy, which is deprecated
```
Comment (mweinelt) [inline pkgs/development/python-modules/django-mptt/default.nix:31]: Set `doCheck = false` in this case.

**Comment by minijackson** [comment]
@xanderio your nixpkgs-review might have failed because this PR depends on the djangorestframework PR. While testing my changes, I had those djangorestframework changes in my local tree.

**Comment by xanderio** [comment]
That explains it `nixpkgs-review` only applied this PR.

**Comment by minijackson** [comment]
Dependent PR was merged, so this is now ready for review!

Both the `netbox_4_0` and the `netbox-upgrade` passes.

---

Result of `nixpkgs-review` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
  </ul>
</details>
<details>
  <summary>22 packages built:</summary>
  <ul>
    <li>disko</li>
    <li>netbox</li>
    <li>python311Packages.django-choices-field</li>
    <li>python311Packages.django-choices-field.dist</li>
    <li>python311Packages.django-htmx</li>
    <li>python311Packages.django-htmx.dist</li>
    <li>python311Packages.django-mptt</li>
    <li>python311Packages.django-mptt.dist</li>
    <li>python311Packages.netbox-reorder-rack</li>
    <li>python311Packages.netbox-reorder-rack.dist</li>
    <li>python311Packages.strawberry-django</li>
    <li>python311Packages.strawberry-django.dist</li>
    <li>python312Packages.django-choices-field</li>
    <li>python312Packages.django-choices-field.dist</li>
    <li>python312Packages.django-htmx</li>
    <li>python312Packages.django-htmx.dist</li>
    <li>python312Packages.django-mptt</li>
    <li>python312Packages.django-mptt.dist</li>
    <li>python312Packages.netbox-reorder-rack</li>
    <li>python312Packages.netbox-reorder-rack.dist</li>
    <li>python312Packages.strawberry-django</li>
    <li>python312Packages.strawberry-django.dist</li>
  </ul>
</details>

**Comment by minijackson** [comment]
I updated it to NetBox 4.0.11, which should be the latest 4.0 release. I'll do 4.1 in another PR, since I got issues with database migrations.

I think this PR is pretty much go to go, then.

File: pkgs/by-name/ne/netbox_4_0/package.nix [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+
+      # Optional dependencies, kept here for backward compatibility
+
+      # for the S3 data source backend
+      boto3
+      # for Git data source backend
+      dulwich
+      # for error reporting
+      sentry-sdk
+    ])
```
Comment (dasJ) [inline pkgs/by-name/ne/netbox_4_0/package.nix:71]: `python-jose` is also needed for OIDC auth.

File: pkgs/development/python-modules/django-choices-field/default.nix [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+  build-system = [ poetry-core ];
+
+  dependencies = [
+    django
+    typing-extensions
+  ];
+
+  nativeCheckInputs = [
+    pytestCheckHook
+    pytest-cov
```
Comment (natsukium) [inline pkgs/development/python-modules/django-choices-field/default.nix:34]:
```suggestion
    pytest-cov-stub
```

File: pkgs/development/python-modules/strawberry-django/default.nix [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+  nativeCheckInputs = [
+    pytestCheckHook
+
+    django-guardian
+    django-mptt
+    django-polymorphic
+    factory-boy
+    pillow
+    psycopg2
+    pytest-cov
```
Comment (natsukium) [inline pkgs/development/python-modules/strawberry-django/default.nix:59]:
```suggestion
    pytest-cov-stub
```

File: pkgs/by-name/ne/netbox_4_0/package.nix (hunk 2 above)
Comment (minijackson) [reply pkgs/by-name/ne/netbox_4_0/package.nix]: Good catch! I've added it through `social-auth-core.passthru.optional-dependencies.openidconnect`

**Comment by minijackson** [comment]
Wait, rebasing now shows a problem with `django-polymorphic`…

**Review by dasJ — APPROVED** [review]

**Comment by dasJ** [comment]
@ofborg eval

**Comment by minijackson** [comment]
Finally found what was the problem! Django was updated from 5.0 to 5.1, due to some vulnerabilities, but Django has breaking changes from 5.0 to 5.1. So I've reintroduced Django 5.0.9 which also has the security fixes (cc @mweinelt).

I think this PR is ready for review (again). I hope to get it before 24.11 releases, as it's already to major versions apart.

**Comment by mweinelt** [comment]
Maybe we can pick https://github.com/netbox-community/netbox/pull/17620 instead of reintroducing django_5_0, which goes EOL a month before NixOS 24.11.

**Comment by AkechiShiro** [comment]
I was not able to run `nixpkgs-review` most likely due to merge conflicts : 
```
...failed to build: Failed to merge aedaa6680aea36a36c3a45d04d0a20704ad11d2c into .../.cache/nixpkgs-review/pr-331750/nixpkgs. git merge failed with exit code 1
```

**Comment by minijackson** [comment]
@AkechiShiro thanks! It should be fixed now. I wish GitHub would notify me when my PRs gets conflicted

**Comment by AkechiShiro** [comment]
Result of `nixpkgs-review pr 331750` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
  </ul>
</details>
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>authentik</li>
  </ul>
</details>
<details>
  <summary>38 packages built:</summary>
  <ul>
    <li>netbox</li>
    <li>peering-manager</li>
    <li>python311Packages.django-choices-field</li>
    <li>python311Packages.django-choices-field.dist</li>
    <li>python311Packages.django-htmx</li>
    <li>python311Packages.django-htmx.dist</li>
    <li>python311Packages.django-mptt</li>
    <li>python311Packages.django-mptt.dist</li>
    <li>python311Packages.django-polymorphic</li>
    <li>python311Packages.django-polymorphic.dist</li>
    <li>python311Packages.django-rest-polymorphic</li>
    <li>python311Packages.django-rest-polymorphic.dist</li>
    <li>python311Packages.drf-spectacular</li>
    <li>python311Packages.drf-spectacular.dist</li>
    <li>python311Packages.netbox-documents</li>
    <li>python311Packages.netbox-documents.dist</li>
    <li>python311Packages.netbox-reorder-rack</li>
    <li>python311Packages.netbox-reorder-rack.dist</li>
    <li>python311Packages.strawberry-django</li>
    <li>python311Packages.strawberry-django.dist</li>
    <li>python312Packages.django-choices-field</li>
    <li>python312Packages.django-choices-field.dist</li>
    <li>python312Packages.django-htmx</li>
    <li>python312Packages.django-htmx.dist</li>
    <li>python312Packages.django-mptt</li>
    <li>python312Packages.django-mptt.dist</li>
    <li>python312Packages.django-polymorphic</li>
    <li>python312Packages.django-polymorphic.dist</li>
    <li>python312Packages.django-rest-polymorphic</li>
    <li>python312Packages.django-rest-polymorphic.dist</li>
    <li>python312Packages.drf-spectacular</li>
    <li>python312Packages.drf-spectacular.dist</li>
    <li>python312Packages.netbox-documents</li>
    <li>python312Packages.netbox-documents.dist</li>
    <li>python312Packages.netbox-reorder-rack</li>
    <li>python312Packages.netbox-reorder-rack.dist</li>
    <li>python312Packages.strawberry-django</li>
    <li>python312Packages.strawberry-django.dist</li>
  </ul>
</details>

**Comment by Mastermindaxe** [comment]
Result of `nixpkgs-review pr 331750` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages marked as broken and skipped:</summary>
  <ul>
    <li>baserow</li>
    <li>baserow.dist</li>
    <li>netbox_3_7</li>
    <li>netbox_4_0</li>
  </ul>
</details>
<details>
  <summary>37 packages built:</summary>
  <ul>
    <li>netbox (netbox_4_1)</li>
    <li>python311Packages.django-choices-field</li>
    <li>python311Packages.django-choices-field.dist</li>
    <li>python311Packages.django-htmx</li>
    <li>python311Packages.django-htmx.dist</li>
    <li>python311Packages.django-mptt</li>
    <li>python311Packages.django-mptt.dist</li>
    <li>python311Packages.django-polymorphic</li>
    <li>python311Packages.django-polymorphic.dist</li>
    <li>python311Packages.django-rest-polymorphic</li>
    <li>python311Packages.django-rest-polymorphic.dist</li>
    <li>python311Packages.drf-spectacular</li>
    <li>python311Packages.drf-spectacular.dist</li>
    <li>python311Packages.netbox-documents</li>
    <li>python311Packages.netbox-documents.dist</li>
    <li>python311Packages.netbox-reorder-rack</li>
    <li>python311Packages.netbox-reorder-rack.dist</li>
    <li>python311Packages.strawberry-django</li>
    <li>python311Packages.strawberry-django.dist</li>
    <li>python312Packages.django-choices-field</li>
    <li>python312Packages.django-choices-field.dist</li>
    <li>python312Packages.django-htmx</li>
    <li>python312Packages.django-htmx.dist</li>
    <li>python312Packages.django-mptt</li>
    <li>python312Packages.django-mptt.dist</li>
    <li>python312Packages.django-polymorphic</li>
    <li>python312Packages.django-polymorphic.dist</li>
    <li>python312Packages.django-rest-polymorphic</li>
    <li>python312Packages.django-rest-polymorphic.dist</li>
    <li>python312Packages.drf-spectacular</li>
    <li>python312Packages.drf-spectacular.dist</li>
    <li>python312Packages.netbox-documents</li>
    <li>python312Packages.netbox-documents.dist</li>
    <li>python312Packages.netbox-reorder-rack</li>
    <li>python312Packages.netbox-reorder-rack.dist</li>
    <li>python312Packages.strawberry-django</li>
    <li>python312Packages.strawberry-django.dist</li>
  </ul>
</details>

File: nixos/modules/services/web-apps/netbox.nix [hunk 5]
```diff
@@ -75,15 +75,19 @@ in {
...
-        if lib.versionAtLeast config.system.stateVersion "24.05"
+        if lib.versionAtLeast config.system.stateVersion "24.11"
+        then pkgs.netbox_4_1
+        else if lib.versionAtLeast config.system.stateVersion "24.05"
         then pkgs.netbox_3_7
         else if lib.versionAtLeast config.system.stateVersion "23.11"
         then pkgs.netbox_3_6
         else if lib.versionAtLeast config.system.stateVersion "23.05"
         then pkgs.netbox_3_5
         else pkgs.netbox_3_3;
```
Comment (SuperSandro2000) [inline nixos/modules/services/web-apps/netbox.nix:86]: We should eventually drop some old versions

Comment (SuperSandro2000) [inline nixos/modules/services/web-apps/netbox.nix:86]:
```suggestion
        else pkgs.netbox_3_6;
```

File: pkgs/development/python-modules/strawberry-django/default.nix [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+    debug-toolbar = [ django-debug-toolbar ];
+    enum = [ django-choices-field ];
+  };
+
+  meta = {
+    description = "Strawberry GraphQL Django extension";
+    homepage = "https://github.com/strawberry-graphql/strawberry-django";
+    changelog = "https://github.com/strawberry-graphql/strawberry-django/blob/${src.rev}/CHANGELOG.md";
+    license = lib.licenses.mit;
+    maintainers = [ ];
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/strawberry-django/default.nix:77]: We should add some maintainers to the new packages

File: pkgs/development/python-modules/django-mptt/default.nix [hunk 7]
```diff
@@ -2,35 +2,34 @@
...
 buildPythonPackage rec {
   pname = "django-mptt";
-  version = "0.13.4";
-  format = "setuptools";
+  version = "0.16";
+  pyproject = true;
 
   src = fetchFromGitHub {
     owner = pname;
     repo = pname;
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/django-mptt/default.nix:17]:
```suggestion
    owner = "django-mptt";
    repo = "django-mptt";
```

File: pkgs/development/python-modules/django-htmx/default.nix [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+    pytest-django
+  ];
+
+  pythonImportsCheck = [ "django_htmx" ];
+
+  meta = {
+    description = "Extensions for using Django with htmx";
+    homepage = "https://github.com/adamchainz/django-htmx";
+    license = lib.licenses.mit;
+    maintainers = [ ];
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/django-htmx/default.nix:42]: same here

File: pkgs/development/python-modules/django-choices-field/default.nix [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+    pytest-django
+  ];
+
+  pythonImportsCheck = [ "django_choices_field" ];
+
+  meta = {
+    description = "Django field that set/get django's new TextChoices/IntegerChoices enum";
+    homepage = "https://github.com/bellini666/django-choices-field";
+    license = lib.licenses.mit;
+    maintainers = [ ];
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/django-choices-field/default.nix:44]: and here

File: pkgs/by-name/ne/netbox_4_1/package.nix [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+in
+py.pkgs.buildPythonApplication rec {
+  pname = "netbox";
+  version = "4.1.3";
+
+  format = "other";
+
+  src = fetchFromGitHub {
+    owner = "netbox-community";
+    repo = pname;
```
Comment (SuperSandro2000) [inline pkgs/by-name/ne/netbox_4_1/package.nix:24]:
```suggestion
    repo = "netbox";
```

File: pkgs/by-name/ne/netbox_4_0/package.nix [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+  src = fetchFromGitHub {
+    owner = "netbox-community";
+    repo = pname;
+    rev = "refs/tags/v${version}";
+    hash = "sha256-0yEz7v5RL1+cqbGDyuyEsywFonJQfPdVIQdL0qLyc04=";
+  };
+
+  patches = [
+    ./custom-static-root.patch
+    # From https://github.com/netbox-community/netbox/pull/17620
```
Comment (SuperSandro2000) [inline pkgs/by-name/ne/netbox_4_0/package.nix:30]: Can't we fetchpatch this?

File: pkgs/development/python-modules/django-mptt/default.nix (hunk 7 above)
Comment (Mastermindaxe) [reply pkgs/development/python-modules/django-mptt/default.nix:17]: Can I ask why this change is needed? I don't understand the need to hardcode this. Trying to get the hang of "why we do things how we do things" 😄

File: pkgs/by-name/ne/netbox_4_0/package.nix (hunk 11 above)
Comment (minijackson) [reply pkgs/by-name/ne/netbox_4_0/package.nix:30]: We can't `fetchpatch` this one, it has conflicts when applied to version 4.0

**Comment by felbinger** [comment]
@minijackson My netbox instance is partly broken now ;( It might be related to something in my setup...
Did you test if you can set a something (e.g. role, platform, site) on an object like a device or vm?
image
image

**Comment by minijackson** [comment]
@felbinger I can reproduce the issue on an instance upgraded from 3.7, but not on a fresh 4.1 instance. Not sure where this comes from yet.

I think the demo instance is recreated from scratch on each upgrade, so this doesn't tell us if the issue is from the packaging or on the NetBox side, unfortunately.

**Comment by minijackson** [comment]
@felbinger Okay, I think I found the issue, and I'll open a fix tomorrow. In the meantime you can run:

```bash
rm -rf /var/lib/netbox/static
netbox-manage collectstatic
```

**Comment by SuperSandro2000** [comment]
We probably want to save the version of netbox somewhere in /var/lib/netbox a d when that changes invalidate the static cache.

Or can we use a cheap symlink?

**Comment by felbinger** [comment]
> @felbinger Okay, I think I found the issue, and I'll open a fix tomorrow. In the meantime you can run:
> 
> ```shell
> rm -rf /var/lib/netbox/static
> netbox-manage collectstatic
> ```
> 
> and hard refresh / clear the cache of the page in your browser.

thanks a lot, it's working!

**Comment by SuperSandro2000** [comment]
or maybe we can build it at runtime like paperless-ngx does https://github.com/NixOS/nixpkgs/blob/7e44cf12936b978af6f37e6ffcfc469bbf40215f/pkgs/applications/office/paperless-ngx/default.nix#L217 ?

**Comment by minijackson** [comment]
> We probably want to save the version of netbox somewhere in /var/lib/netbox a d when that changes invalidate the static cache.

We're actually already doing that here: https://github.com/NixOS/nixpkgs/blob/71408df70fd5bfbc0e5fc3776a942e2b0b93dad6/nixos/modules/services/web-apps/netbox.nix#L301-L303

The issue was that old files were not cleared, and it seems upstream's `upgrade.sh` script doesn't do it either: https://github.com/netbox-community/netbox/blob/58bc3884578a98595b02d7d7412c83d41fc7bbfa/upgrade.sh#L100-L101

> Or can we use a cheap symlink?

That's a great suggestion, I was thinking of putting the hash in order for the upgrade script to be run when plugins are added, but the symlink is probably even simpler. Thanks!

> or maybe we can build it at runtime like paperless-ngx does

We're already doing that, but I actually didn't know of the `--clear` option, which is exactly what I was looking for!

**Pull request merged**